feat: Add instrumentation for tracking variables in debugging

ggiraldez opened the pull request:
# Description

## Problem\*

Part of #3015 

## Summary\*

This PR injects instrumentation snippets to the compiled code when running under a debugger to track the values assigned to variables in the program. It also provides the debugging context with necessary support code (in the form of foreign functions) and new functionality to inspect the tracked values.

Instrumentation occurs in two phases:

1. During parsing, when assignments are detected, they are replaced by a small snippet that captures the value assigned and a temporary identifier for the variable being assigned, and a foreign function (oracle) is called with this information.
2. At monomorphization time, ie. after the exact types of all variables has been determined, replacing the temporary identifier by a final one which will depend on the variable itself and the type at each instance of the function being compiled (for generic functions). Also, since structs are replaced for tuples during compilation, at this stage field and other member accesses is resolved for the final types (ie. indices in the tuples).

## Additional Context

Besides the runtime support in the debugger, the compiler also injects a synthetic crate `__debug` which holds the definition of the oracle functions used in the injected code for tracking the variables and their assigned values.

These changes were extracted from the `dap-with-vars` branch of `manastech/noir` repository where most of the development of this feature occurred. We will submit additional PRs for the remaining features in that branch, notably improved REPL and DAP support to make use of this instrumentation code.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [X] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

ggiraldez commented:
Failures in `wasm/compiler` are because we're injecting the `__debug` crate always, regardless of whether we're debugging or not. I'll work on addressing this issue and update the PR.

mverzilli commented:
Fixes #4026

TomAFrench commented on compiler/noirc_driver/src/lib.rs:
```diff
@@ -136,6 +155,17 @@ pub fn prepare_crate(context: &mut Context, file_name: &Path) -> CrateId {
 }
 
+pub fn link_to_debug_crate(context: &mut Context, root_crate_id: CrateId) {
+    let path_to_debug_lib_file = Path::new(DEBUG_CRATE_NAME).join("lib.nr");
+    let debug_file_id = context
+        .file_manager
+        .name_to_id(path_to_debug_lib_file)
+        .expect("debug source is expected to be present in file manager");
+    let debug_crate_id = context.crate_graph.add_crate(debug_file_id);
+
+    add_dep(context, root_crate_id, debug_crate_id, DEBUG_CRATE_NAME.parse().unwrap());
```
Looks like this can be replaced with a call to `prepare_dependency` and `add_dep` to avoid duplication.

ggiraldez replied:
That will also link the debug crate with stdlib. I guess it doesn't hurt. I'll try replacing it.

TomAFrench commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,610 @@
+            scope: vec![],
+            next_var_id: 0,
+            next_field_name_id: 1,
+            enabled: true, // TODO
```
Unresolved TODO.

ggiraldez replied:
Removed the field altogether since it was not being used.

TomAFrench commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,610 @@
+
+        self.walk_scope(&mut f.body.0);
+
+        // prapend fn params:
```
```suggestion
        // prepend fn params:
```

TomAFrench replied:
Forgot I can't push these changes myself. Can deal with these later.

TomAFrench commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,610 @@
+            }
+            ast::LValue::Dereference(_lv) => {
+                // TODO: this is a dummy statement for now, but we should
+                // somehow track the derefence and update the pointed to
```
```suggestion
                // somehow track the dereference and update the pointed to
```

TomAFrench commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,610 @@
+    }
+}
+
+fn none_span() -> Span {
+    Span::build_from_str("")
+}
```
Replace with `Span::empty()`?

ggiraldez replied:
Done!

TomAFrench commented on compiler/noirc_errors/src/position.rs:
```diff
@@ -86,6 +86,10 @@ impl Span {
         self.0.end().into()
     }
 
+    pub fn build_from_str(s: &str) -> Span {
+        Span(ByteSpan::from_str(s))
+    }
+
```
Which would make this unused.

TomAFrench commented on tooling/nargo_cli/src/cli/debug_cmd.rs:
```diff
@@ -46,6 +49,10 @@ pub(crate) fn run(
     args: DebugCommand,
     config: NargoConfig,
 ) -> Result<(), CliError> {
+    // TODO: set clap default for flattened flag
```
Unresolved TODO

ggiraldez replied:
This was a leftover.

TomAFrench reviewed (commented):
I'm going to let Jake take the first look at the core of this PR. I've made a couple of notes around the edges.

ggiraldez commented on compiler/fm/src/lib.rs:
```diff
@@ -78,6 +78,21 @@ impl FileManager {
         Some(file_id)
     }
 
+    pub fn add_file_with_contents(&mut self, file_name: &Path, contents: &str) -> Option<FileId> {
+        // Handle both relative file paths and std/lib virtual paths.
+        let resolved_path: PathBuf = file_name.to_path_buf();
+
+        // Check that the resolved path already exists in the file map, if it is, we return it.
+        if let Some(file_id) = self.path_to_id.get(&resolved_path) {
+            return Some(*file_id);
+        }
+
+        // Otherwise we add the file
+        let file_id = self.file_map.add_file(resolved_path.clone().into(), String::from(contents));
+        self.register_path(file_id, resolved_path);
+        Some(file_id)
+    }
```
Oh! It seems that we [originally implemented](https://github.com/manastech/noir/commit/e5d5bc889fb4d917db1c09fe69972c0a437395b4) that a couple of days before `add_file_with_source_canonical_path`. I'll remove it. Thanks!

ggiraldez commented on tooling/nargo/src/artifacts/debug.rs:
```diff
@@ -23,22 +24,15 @@ impl DebugArtifact {
-                    .flat_map(|call_stack| call_stack.iter().map(|location| location.file))
-            })
-            .collect();
+        let file_ids: BTreeSet<FileId> =
+            debug_symbols.iter().flat_map(|debug_info| debug_info.get_file_ids()).collect();
```
I'm not sure why this change was there, but it seems it was mostly doing the same thing as the current code. Probably a quasi simultaneous but parallel change again. I'm removing the change.

jfecher commented on compiler/noirc_frontend/src/monomorphization/debug_types.rs:
```diff
@@ -0,0 +1,70 @@
+        });
+        let existing_var_id = self.fe_to_vars.get(&fe_var_id).and_then(|var_id| {
+            let (_, existing_type_id) = self.variables.get(var_id).unwrap();
+            if *existing_type_id == type_id {
+                Some(var_id)
+            } else {
+                None
+            }
```
Nitpick:
```suggestion
            (*existing_type_id == type_id).then_some(var_id)
```

ggiraldez replied:
Nice!

jfecher commented on compiler/noirc_frontend/src/monomorphization/debug_types.rs:
```diff
@@ -0,0 +1,70 @@
+            type_id
+        });
+        let existing_var_id = self.fe_to_vars.get(&fe_var_id).and_then(|var_id| {
+            let (_, existing_type_id) = self.variables.get(var_id).unwrap();
```
Is this guaranteed `var_id` is present in this hash map? I can't seem to tell from the function

ggiraldez replied:
Yes, it's guaranteed because if the `fe_var_id` is found, it means it's already being mapped to at least one specific (ie. non-generic) data type and instantiated in the tracked variables map. The code for this insertion is in the else part at the end of this method.

I do agree it's confusing. I'll replace the `u32` types with more specific ones and re-evaluate the resulting code.

There is also one "hidden" assumption in all this code which I don't think we documented properly. Given an `fe_var_id` we don't always have access to the underlying type (eg. for dropping them), and so we don't know if we're using the "correct" `var_id`. The assumption is that all references to the variable are local to the function being monomorphized and they happen *after* the variable is declared or assigned to, and we *do* have the type at that precise moment. When that happens we either reuse the `var_id` or we create a fresh new one, which should be correct for the rest of the scope being compiled. TBH, we didn't initially take into account generic functions, so we introduced the mapping from `fe_var_id` into `var_id` later in the development.

ggiraldez replied:
I added several comments in the code explaining the difference between the `source_var_id` and the `var_id`. There's also the description of the `source_to_debug_vars` field in `DebugTypeTracker` that explains why the mapping is necessary.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -224,30 +255,43 @@ impl<'interner> Monomorphizer<'interner> {
         main_meta.function_signature()
     }
 
+    fn compile_main_debug(
+        &mut self,
+        main_id: node_interner::FuncId,
+        field_names: &HashMap<u32, String>,
+    ) -> FunctionSignature {
+        self.debug_field_names = field_names.clone();
+        self.compile_main(main_id)
+    }
+
```
```suggestion
```
We can remove this function and just pass `field_names` to Monomorphizer's constructor.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -93,9 +100,29 @@ type HirType = crate::Type;
+pub fn monomorphize_debug(
+    main: node_interner::FuncId,
+    interner: &mut NodeInterner,
+    field_names: &HashMap<u32, String>,
```
```suggestion
    field_names: HashMap<u32, String>,
```
We should take in `field_names` by value since we need to clone it anyway. That way if a caller can move in the map, we avoid a clone.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -224,30 +255,43 @@ impl<'interner> Monomorphizer<'interner> {
-            Type::TraitAsType(..) => &body_return_type,
-            _ => meta.return_type(),
-        });
-
-        let parameters = self.parameters(&meta.parameters);
+            let body_expr_id = *self.interner.function(&f).as_expr();
+            let body_return_type = self.interner.id_type(body_expr_id);
+            let return_type = self.convert_type(match meta.return_type() {
+                Type::TraitAsType(_, _, _) => &body_return_type,
+                _ => meta.return_type(),
+            });
+            let unconstrained = modifiers.is_unconstrained
+                || matches!(modifiers.contract_function_type, Some(ContractFunctionType::Open));
+            (meta.parameters, unconstrained, body_expr_id, name, return_type)
+        };
 
-        let body = self.expr(body_expr_id);
-        let unconstrained = modifiers.is_unconstrained
-            || matches!(modifiers.contract_function_type, Some(ContractFunctionType::Open));
+        let function = {
+            let parameters = self.parameters(&meta_parameters).clone();
+            let body = self.expr(body_expr_id).clone();
+            ast::Function { id, name, parameters, body, return_type, unconstrained }
+        };
```
What changed about this section? It looks like we're introducing a lot more `clone`s but I don't see changes beyond that. I know the interner is mutable now but I think we should be able to avoid most of these clones still.

ggiraldez replied:
You're right. We only need to `clone()` the `meta` variable. The rest can remain the same. Maybe something changed, but I distinctly remember fighting the borrow checker on this one. But maybe it was just my inexperience. Thanks!

ggiraldez replied:
The other required change was in the order of some of the statements, due to borrow checker constraints.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -909,6 +953,142 @@ impl<'interner> Monomorphizer<'interner> {
+    /// For this last part, we need to resolve the mapping from field names in
+    /// structs to positions in the runtime tuple, since all structs are
+    /// replaced by tuples during compilation.
+    fn patch_debug_member_assign(
+        &mut self,
+        call: &HirCallExpression,
+        arguments: &mut [ast::Expression],
+        arity: usize,
+    ) {
```
Can we move these `patch_*` functions to their own dedicated file? E.g. perhaps we should have a monomorphization/debug module.

ggiraldez replied:
Yes, that's a good idea.

ggiraldez replied:
All code (mostly) in the monomorphization phase related to debugging now lives in separate files `debug.rs` and `debug_types.rs`.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -909,6 +953,142 @@ impl<'interner> Monomorphizer<'interner> {
+            .clone();
+        for i in 0..arity {
+            if let Some(HirExpression::Literal(HirLiteral::Integer(fe_i, i_neg))) =
+                hir_arguments.get(2 + i)
```
2?
Just parsing this quickly I'm unsure what the offset is for. Can you use a named constant instead?

ggiraldez replied:
I introduced constants to name the argument slots for the debug oracle functions. 2 was the indeed the position of the first index to reference member assignments (right after the var_id and the updated value for the member).

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -909,6 +953,142 @@ impl<'interner> Monomorphizer<'interner> {
+            }
+        }
+
+        let var_id = if &var_def_name != "__debug_expr" {
```
This check seems a bit fragile

ggiraldez replied:
Agreed, and this actually shouldn't be necessary because when doing member assignment, the variable should be already instantiated for the current type and we will always assign from `__debug_expr`. I'm thinking this logic is a leftover from a previous implementation. In fact, we should be able to capture the names of the variables during injection and do the instantiation/lookup by type here without needing the name.

ggiraldez replied:
This bit of logic changed in the last commits. The name of the variables are now carried over from the instrumentation phase, so there's no need to check of the value passed into the oracle is the variable itself or the temporary variable `__debug_expr`. This was brittle as you said.

jfecher commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -1590,3 +1783,30 @@ fn undo_instantiation_bindings(bindings: TypeBindings) {
+    }
+}
+
+fn next_type(ptype: &PrintableType, i: usize) -> PrintableType {
+    match ptype {
+        PrintableType::Array { length: _length, typ } => (**typ).clone(),
+        PrintableType::Tuple { types } => types[i].clone(),
+        PrintableType::Struct { name: _name, fields } => fields[i].1.clone(),
+        PrintableType::String { length: _length } => {
+            // TODO: check bounds
+            PrintableType::UnsignedInteger { width: 8 }
+        }
+        _ => {
+            panic!["expected type with sub-fields, found terminal type"]
+        }
+    }
+}
```
I think `element_type_at_index` would be a better name here. It also looks like this `TODO` wouldn't be terribly difficult to fill in. I also think it's fine to remove though since the only element type of a String would be a character anyway.

ggiraldez replied:
I think this refers to checking that the `i` index accessed is within the length of the string. This is needed in the context of assigning to elements of a string, which I don't think is possible in Noir, correct?

I'm not sure it's a known issue, but I get a weird error when trying to do:
```noir
fn main() {
    let mut s = "hello";
    s[0] = s[1];
}
```
Here's the error:
```
error: Expected type array, found type str<5>
  ┌─ /home/ggiraldez/Scratchpad/mutable_str/src/main.nr:3:13
  │
3 │     s[10] = 64;
  │             --
  │
```

jfecher replied:
Hmm, the location of that error is wrong. It's trying to say it expected `s` to be an array to be accessed by index like that. This could be a good issue to file. Currently strings don't support many operations at all and are mostly for debug printing purposes.

jfecher replied:
> I think this refers to checking that the i index accessed is within the length of the string. This is needed in the context of assigning to elements of a string, which I don't think is possible in Noir, correct?

I don't think it should be required for `next_type` specifically though since the only type for strings is the character type. Similar to the array case where we don't check the size either.

ggiraldez replied:
Agreed and reflected in the last commits. I'll file an issue later on the error location.

jfecher commented on tooling/nargo/src/artifacts/debug_vars.rs:
```diff
@@ -0,0 +1,132 @@
+pub struct DebugVars {
+    id_to_name: HashMap<u32, String>,
+    active: HashSet<u32>,
+    id_to_value: HashMap<u32, PrintableValue>, // TODO: something more sophisticated for lexical levels
```
A Vec of scopes perhaps?

ggiraldez replied:
Yes, lexical scopes are handled in #4188. But I see we haven't put a proper name for this there either. I'll change this along with changing the `u32` to more descriptive type names.

ggiraldez replied:
This was changed to use proper identifier types. Lexical scope handling is left for #4188.

jfecher commented on tooling/nargo/src/artifacts/debug_vars.rs:
```diff
@@ -0,0 +1,132 @@
+        self.id_to_type.get(&var_id).and_then(|type_id| self.types.get(type_id))
+    }
+
+    pub fn drop(&mut self, var_id: u32) {
```
Can you rename this to `drop_var` or `remove_var`? When I read `drop` it sounds like we are implementing `Drop` and dropping the entire struct.

jfecher reviewed (commented):
I think this PR would benefit greatly from adding ID types for its variables, types, and fe_vars which all currently use `u32` in the source code right now. This makes it hard to tell what a `u32` is for without documentation and potentially easy to use a fe_var u32 where a type was expected or vice-versa.

jfecher reviewed (changes requested):
Meant to request changes

ggiraldez commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,602 @@
+            })
+            .collect();
+
+        let set_fn_params = pvars
+            .iter()
+            .map(|(var_id, id, _is_mut)| self.wrap_assign_var(*var_id, id_expr(id)))
+            .collect();
```
Simplified this code and changed into a single iteration.

ggiraldez commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,602 @@
+
+        statements.iter_mut().for_each(|stmt| self.walk_statement(stmt));
+
+        let (ret_stmt, fn_body) =
+            statements.split_last().map(|(e, b)| (e.clone(), b.to_vec())).unwrap_or((
+                ast::Statement {
+                    kind: ast::StatementKind::Expression(ast::Expression {
+                        kind: ast::ExpressionKind::Literal(ast::Literal::Unit),
+                        span: end_span,
+                    }),
+                    span: end_span,
+                },
+                vec![],
+            ));
```
You're right! I basically rewrote most of this function, following a more imperative stye. Hopefully the intent is more clear now and it's more efficient as well.

ggiraldez commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,602 @@
+            // assign return expr to __debug_expr:
+            vec![match &ret_stmt.kind {
+                ast::StatementKind::Expression(ret_expr) => ast::Statement {
+                    kind: ast::StatementKind::Let(ast::LetStatement {
+                        pattern: ast::Pattern::Identifier(ident("__debug_expr", ret_expr.span)),
+                        r#type: ast::UnresolvedType::unspecified(),
+                        expression: ret_expr.clone(),
+                    }),
+                    span: ret_expr.span,
+                },
+                _ => ret_stmt.clone(),
+            }],
+            // drop fn params:
+            self.scope
+                .pop()
+                .unwrap_or(HashMap::default())
+                .values()
+                .map(|var_id| self.wrap_drop_var(*var_id, end_span))
+                .collect(),
+            // return the __debug_expr value:
+            vec![match &ret_stmt.kind {
+                ast::StatementKind::Expression(_ret_expr) => ast::Statement {
+                    kind: ast::StatementKind::Expression(ast::Expression {
+                        kind: ast::ExpressionKind::Variable(ast::Path {
+                            segments: vec![ident("__debug_expr", ret_stmt.span)],
+                            kind: PathKind::Plain,
+                            span: ret_stmt.span,
+                        }),
+                        span: ret_stmt.span,
+                    }),
+                    span: ret_stmt.span,
+                },
+                _ => ast::Statement {
+                    kind: ast::StatementKind::Expression(ast::Expression {
+                        kind: ast::ExpressionKind::Literal(ast::Literal::Unit),
+                        span: ret_stmt.span,
+                    }),
+                    span: ret_stmt.span,
+                },
+            }],
```
Yes, absolutely! See my previous comment.

ggiraldez commented on compiler/noirc_frontend/src/hir/mod.rs:
```diff
@@ -41,6 +44,8 @@ pub struct Context<'file_manager, 'parsed_files> {
     pub parsed_files: Cow<'parsed_files, ParsedFiles>,
 }
 
+pub type StorageSlot = u32;
+
```
I think this is a leftover from a git merge. I'll remove it.

ggiraldez commented on compiler/noirc_frontend/src/monomorphization/debug_types.rs:
```diff
@@ -0,0 +1,70 @@
+use noirc_printable_type::PrintableType;
+use std::collections::HashMap;
+
+/// We keep a collection of the debug variables and their types in this
+/// structure. The fe_var_id refers to the ID given when inserting the
+/// instrumentation probe. This variable does not have a type yet and hence it
+/// can be instantiated multiple types if it's in the context of a generic
+/// variable. The var_id refers to the ID of the instantiated variable which
+/// will have a valid type.
+#[derive(Debug, Clone, Default)]
+pub struct DebugTypes {
+    fe_to_vars: HashMap<u32, u32>,          // fe_var_id => var_id
+    variables: HashMap<u32, (String, u32)>, // var_id => (var_name, type_id)
+    types: HashMap<PrintableType, u32>,
+    id_to_type: HashMap<u32, PrintableType>,
+    next_type_id: u32,
+    next_var_id: u32,
+}
```
Oh, I see how this is very confusing. `fe_var_id` refers to the ID of variables inserted in the frontend, ie. during the instrumentation injection. But this is not entirely correct since monomorphization is still part of the frontend. Long story short, this is different from the `var_id` because for generic functions we will end up with multiple instances of the same variable, all with different types. Ie. one `fe_var_id` corresponds to one or more `var_id`. `var_id` is the instantiated variable for which we keep track of the value. I'll think of better names.

ggiraldez replied:
While changing all `u32` to more appropriate types, I renamed quite a few things. `fe_var_id` is now called `source_var_id` to convey the fact that it refers to a variable found in the source code, while `var_id` (or `debug_var_id`) is used to name the runtime variable which has a unique type, determined by the monomorphizer.

ggiraldez commented on compiler/noirc_errors/src/debug_info.rs:
```diff
@@ -28,6 +33,8 @@ pub struct DebugInfo {
     #[serde_as(as = "BTreeMap<DisplayFromStr, _>")]
     pub locations: BTreeMap<OpcodeLocation, Vec<Location>>,
+    pub variables: HashMap<u32, (String, u32)>, // var_id => (name, type_id)
+    pub types: HashMap<u32, PrintableType>,     // type_id => printable type
```
All identifier types are now tuple structs and this should make the code more robust and easier to follow.

ggiraldez commented on compiler/noirc_frontend/src/debug/mod.rs:
```diff
@@ -0,0 +1,602 @@
+    fn walk_fn(&mut self, f: &mut ast::FunctionDefinition) {
+        self.scope.push(HashMap::default());
+
+        let pvars: Vec<(u32, ast::Ident, bool)> = f
```
This code was simplified a bit and `pvars` no longer exists.

ggiraldez commented:
> I think this PR would benefit greatly from adding ID types for its variables, types, and fe_vars which all currently use `u32` in the source code right now. This makes it hard to tell what a `u32` is for without documentation and potentially easy to use a fe_var u32 where a type was expected or vice-versa.

I addressed all comments and made a bunch of changes. I did some renames as well for some of the types and variables that we introduced for the instrumentation. I think the code reads a lot better now and is easier to follow.

jfecher reviewed (approved):
Thanks for all the updates, LGTM :+1:

jfecher commented on compiler/noirc_frontend/src/monomorphization/debug_types.rs:
```diff
@@ -0,0 +1,70 @@
+    fe_to_vars: HashMap<u32, u32>,          // fe_var_id => var_id
+    variables: HashMap<u32, (String, u32)>, // var_id => (var_name, type_id)
+    types: HashMap<PrintableType, u32>,
+    id_to_type: HashMap<u32, PrintableType>,
```
I'm not sure. It is just odd seeing `PrintableType`s used for purposes outside of printing. It may be something to mark to change / re-examine for later. `hir_def::Type` should always be fully resolved since it is the result of resolving an `UnresolvedType`. If you meant any type variables should be followed, there is `.follow_bindings()` to follow all Bound type variable links if that helps. We should also be able to make types serializable as well but I'll leave the ultimate decision to you here.

TomAFrench commented on compiler/noirc_driver/src/lib.rs:
```diff
@@ -83,6 +85,14 @@ pub struct CompileOptions {
     pub show_monomorphized: bool,
+
+    /// Insert debug symbols to inspect variables
+    #[arg(long)]
```
```suggestion
    #[arg(long, hide = true)]
```
